packages/flutter_tools/test/general.shard/flutter_tester_device_test.dart

```diff
@@ -50,6 +50,18 @@ void main() {
       uriConverter: (String input) => '$input/converted',
     );
 
+  testUsingContext('Missing dir error caught for FontConfigManger.dispose', () async {
+    final FontConfigManager fontConfigManager = FontConfigManager();
+
+    final Directory fontsDirectory = fileSystem.file(fontConfigManager.fontConfigFile).parent;
+    fontsDirectory.deleteSync(recursive: true);
+
+    expect(fontConfigManager.dispose, returnsNormally);
```

Comment:
`returnsNormally` is not an async matcher, so this is testing that `dispose` returns a future without throwing, not that the future does not complete with an error. why don't you just call the function and await the result without wrapping it in an expect (if it throws, the test will still fail).

Reply:
exceptions won't crash the test runner, but will be treated as passes.

Reply:
Hmm maybe I'm messing up something but I'm trying to get the rest of the tests to run after purposefully throwing an `Exception` within the dispose method, I have edited it to have the below

```dart
  Future<void> dispose() async {
    if (_fontsDirectory != null) {
      globals.printTrace('Deleting ${_fontsDirectory!.path}...');
      try {
        await _fontsDirectory!.delete(recursive: true);
      } on FileSystemException {
        // Silently exit
      }
      _fontsDirectory = null;
      throw Exception('error'); // Purposefully thrown
    }
  }
```

When I try to run the test file with `flutter test test/general.shard/flutter_tester_device_test.dart`, the process hangs after the first uncaught exception.

Same result for `dart test test/general.shard/flutter_tester_device_test.dart`

However, with the `expect(fontConfigManager.dispose, returnsNormally);`, it does highlight the error and it continues with the rest of the tests in file.  Am I missing something? The logs below show the output from the test runner when using the `returnsNormally` using `flutter test`

```
eliasyishak-macbookpro:flutter_tools eliasyishak$ flutter test /Users/eliasyishak/Desktop/flutter/packages/flutter_tools/test/general.shard/flutter_tester_device_test.dart
00:00 +0: Missing dir error caught for FontConfigManger.dispose                                                                                                                                                                            
Invalid argument(s): error
#0      FontConfigManager.dispose (package:flutter_tools/src/test/font_config_manager.dart:43:7)
<asynchronous suspension>

00:01 +0 -1: Missing dir error caught for FontConfigManger.dispose [E]                                                                                                                                                                     
  Invalid argument(s): error
  package:flutter_tools/src/test/font_config_manager.dart 43:7  FontConfigManager.dispose
  ===== asynchronous gap ===========================
  dart:async                                                    _CustomZone.registerBinaryCallback
  package:flutter_tools/src/test/font_config_manager.dart 38:9  FontConfigManager.dispose
  package:matcher                                               expect
  test/general.shard/flutter_tester_device_test.dart 59:5       main.<fn>
  test/src/context.dart 141:42                                  testUsingContext.<fn>.<fn>.<fn>.<fn>.<fn>
  package:flutter_tools/src/base/context.dart 150:29            AppContext.run.<fn>
  dart:async                                                    runZoned
  package:flutter_tools/src/base/context.dart 149:12            AppContext.run
  test/src/context.dart 130:30                                  testUsingContext.<fn>.<fn>.<fn>.<fn>
  dart:async                                                    runZonedGuarded
  test/src/context.dart 128:18                                  testUsingContext.<fn>.<fn>.<fn>
  package:flutter_tools/src/base/context.dart 150:29            AppContext.run.<fn>
  dart:async                                                    runZoned
  package:flutter_tools/src/base/context.dart 149:12            AppContext.run
  test/src/context.dart 103:22                                  testUsingContext.<fn>.<fn>
  package:flutter_tools/src/context_runner.dart 84:18           runInContext.runnerWrapper
  ===== asynchronous gap ===========================
  dart:async                                                    _CustomZone.registerUnaryCallback
  package:flutter_tools/src/context_runner.dart 83:20           runInContext.runnerWrapper
  package:flutter_tools/src/base/context.dart 150:29            AppContext.run.<fn>
  dart:async                                                    runZoned
  package:flutter_tools/src/base/context.dart 149:12            AppContext.run
  package:flutter_tools/src/context_runner.dart 87:18           runInContext
  test/src/context.dart 102:11                                  testUsingContext.<fn>
  test/src/common.dart 183:18                                   test.<fn>
  

To run this test again: /Users/eliasyishak/Desktop/flutter/bin/cache/dart-sdk/bin/dart test /Users/eliasyishak/Desktop/flutter/packages/flutter_tools/test/general.shard/flutter_tester_device_test.dart -p vm --plain-name 'Missing dir error caught for FontConfigManger.dispose'
00:01 +7 -1: Some tests failed.                                                                                                                                                                                                            

Oops; flutter has exited unexpectedly: "FileSystemException: error, path = ''".
^C
Oops; flutter has exited unexpectedly: "FileSystemException: error, path = ''".
```

Reply:
For both approaches, I have to `^C` out of each process though....

Reply:
Yeah, but try putting:

```dart
        await Future<void>.delayed(const Duration(seconds: 10));
```

before your exception...

Reply:
(for me the test passes, which means whether or not this test passes or fails is a race condition)

Reply:
Hmm interesting, I can go ahead and just await the result though, I was just going down the rabbit hole

Reply:
I think the fact that uncaught exceptions hang the runner from running other tests is a bug with how we manage zones/context. consider this test:

```dart
import '../src/common.dart';

void main() {
  test('1', () => throw Exception('oops'));
  test('2', () {});
  test('3', () {});
}
```

When I run this single threaded, it runs all tests, just failing the first, without hanging the runner at all.

Reply:
But if I substitute `testUsingContext`, the first hangs and the rest of the tests don't run.

Reply:
> But if I substitute testUsingContext

Ah yep, thats it, I was assuming it would behave the same as `test`. That helps, thanks for the breakdown

Reply:
This was good we narrowed it down to testUsingContext--we should fix this :)

Reply:
Yeah I'll create an issue for this incase we don't have it already